model/labels/labels.go

```diff
@@ -205,73 +216,71 @@ func (ls Labels) HashWithoutLabels(b []byte, names ...string) (uint64, []byte) {
 // BytesWithLabels is just as Bytes(), but only for labels matching names.
 // 'names' have to be sorted in ascending order.
 func (ls Labels) BytesWithLabels(buf []byte, names ...string) []byte {
-	b := bytes.NewBuffer(buf[:0])
-	b.WriteByte(labelSep)
-	i, j := 0, 0
-	for i < len(ls) && j < len(names) {
-		if names[j] < ls[i].Name {
-			j++
-		} else if ls[i].Name < names[j] {
-			i++
-		} else {
-			if b.Len() > 1 {
-				b.WriteByte(seps[0])
-			}
-			b.WriteString(ls[i].Name)
-			b.WriteByte(seps[0])
-			b.WriteString(ls[i].Value)
-			i++
+	b := buf[:0]
+	j := 0
+	for pos := 0; pos < len(ls.data); {
+		lName, newPos := decodeString(ls.data, pos)
+		_, newPos = decodeString(ls.data, newPos)
+		for j < len(names) && names[j] < lName {
 			j++
 		}
+		if j == len(names) {
+			break
+		}
+		if lName == names[j] {
+			b = append(b, ls.data[pos:newPos]...)
+		}
+		pos = newPos
 	}
-	return b.Bytes()
+	return b
 }
 
 // BytesWithoutLabels is just as Bytes(), but only for labels not matching names.
 // 'names' have to be sorted in ascending order.
 func (ls Labels) BytesWithoutLabels(buf []byte, names ...string) []byte {
-	b := bytes.NewBuffer(buf[:0])
-	b.WriteByte(labelSep)
+	b := buf[:0]
 	j := 0
-	for i := range ls {
-		for j < len(names) && names[j] < ls[i].Name {
+	for pos := 0; pos < len(ls.data); {
+		lName, newPos := decodeString(ls.data, pos)
+		_, newPos = decodeString(ls.data, newPos)
+		for j < len(names) && names[j] < lName {
 			j++
 		}
-		if j < len(names) && ls[i].Name == names[j] {
-			continue
+		if j == len(names) || lName != names[j] {
+			b = append(b, ls.data[pos:newPos]...)
 		}
-		if b.Len() > 1 {
-			b.WriteByte(seps[0])
-		}
-		b.WriteString(ls[i].Name)
-		b.WriteByte(seps[0])
-		b.WriteString(ls[i].Value)
+		pos = newPos
 	}
-	return b.Bytes()
+	return b
 }
 
 // Copy returns a copy of the labels.
 func (ls Labels) Copy() Labels {
-	res := make(Labels, len(ls))
-	copy(res, ls)
-	return res
+	buf := append([]byte{}, ls.data...)
```

Comment:
Since strings are immutable is this really needed?
I would think that `Labels{data: ls.data}` would be enough and there's no need to allocate anything.

Reply:
Strings are immutable except when people use `unsafe`...

It is used in places like this (on current `main`): https://github.com/prometheus/prometheus/blob/3947238ce0e3ff7e7c6e9502f73275cdbc233054/tsdb/querier.go#L509-L510

which I replaced with a call to `Labels.Copy()`.

To expand a little, we are re-using the buffer inside `labels.Labels` to reduce memory allocations.
This is non-`unsafe` when you do it with slices, but you still have to be extremely careful to copy in all the right places.

Probably there is a cleaner way where we pass some higher-level iterator object around.